runtime/lua/vim/lsp/folding_range.lua

```diff
@@ -0,0 +1,320 @@
+local util = require('vim.lsp.util')
+local log = require('vim.lsp.log')
+local ms = require('vim.lsp.protocol').Methods
+local api = vim.api
+
+local M = {}
+
+---@class (private) vim.lsp.folding_range.BufState
+---
+---@field version? integer
+---
+--- Index in the form of row -> foldlevel
+---@field row_level table<integer, [integer, nil | ">" | "<"]>
+---
+--- Index in the form of start_row -> collapsed_text
+---@field row_text table<integer, string>
+---
+--- Index in the form of kind -> range
+---@field kind_ranges table<lsp.FoldingRangeKind, lsp.FoldingRange[]?>>
+
+---@type table<integer, vim.lsp.folding_range.BufState?>
+local bufstates = {}
+
+--- Update visible folds in the buffer.
+---@param bufnr integer
+local function foldupdate(bufnr)
+  for _, winid in ipairs(vim.fn.win_findbuf(bufnr)) do
+    local wininfo = vim.fn.getwininfo(winid)[1]
+    if wininfo and wininfo.tabnr == vim.fn.tabpagenr() then
+      if vim.wo[winid].foldmethod == 'expr' then
+        vim._foldupdate(winid, 0, api.nvim_buf_line_count(bufnr))
+      end
+    end
+  end
+end
+
+local augroup_update = api.nvim_create_augroup('vim_lsp_folding_range/update', {})
+
+--- Schedule `foldupdate()` after leaving insert mode.
+---@param bufnr integer
+local function schedule_foldupdate(bufnr)
+  if #(api.nvim_get_autocmds({
+    group = augroup_update,
+    buffer = bufnr,
+  })) == 0 then
```

Comment:
Instead of collecting autocmds, would it be better to send an id and check for that instead?

Reply:
I have changed it to a method very similar to your suggestion, and now I will mark this thread as resolved.